apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,188 @@
+// idempotency.interceptor.ts
+
+import {
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+  Logger,
+  ConflictException,
+  HttpException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { CacheService } from '@novu/application-generic';
+import { Observable, of, throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+import { createHash } from 'crypto';
+
+const LOG_CONTEXT = 'IdempotencyInterceptor';
+const IDEMPOTENCY_CACHE_TTL = 60 * 60 * 6; //6h
+const IDEMPOTENCY_PROGRESS_TTL = 60 * 5; //5min
+
+const HEADER_KEYS = {
+  IDEMPOTENCY: 'idempotency-key',
```

Comment:
I think the header keys should be capitalize i.e `Idempotency-Key` just to enforce consistency

Reply:
enforce consistency of what exactly?

Reply:
@Cliftonz I mean most times response headers' first letter is always capitalized. i.e. `idempotency-key` can be `Idempotency-Key`.

Reply:
Yeah, Pascal-Case is the convention. I am good with this.

Reply:
by default headers are lowercased by the time it reaches the interceptor, so trying to access using Pascal-Case key returns undefined value. I have tried that and changed to lowercase.

Reply:
@mahendraHegde Thanks for verifying this!